spec/abci++/abci++_methods_002_draft.md

```diff
@@ -290,13 +290,11 @@ title: Methods
     | hash                    | bytes                                       | The block header's hash of the block to propose. Present for convenience (can be derived from the block header). | 1            |
     | header                  | [Header](../core/data_structures.md#header) | The header of the block to propose.                                                                              | 2            |
     | txs                     | repeated bytes                              | Preliminary list of transactions that have been picked as part of the block to propose.                          | 3            |
-    | last_commit_info        | [LastCommitInfo](#lastcommitinfo)           | Info about the last commit, including the round, the validator list, and which ones signed the last block.       | 4            |
+    | local_last_commit       | [ExtendedCommitInfo](#extendedcommitinfo)   | Info about the last commit, obtained locally from Tendermint's data structures.                                  | 4            |
     | byzantine_validators    | repeated [Evidence](#evidence)              | List of evidence of validators that acted maliciously.                                                           | 5            |
     | max_tx_bytes            | int64                                       | Currently configured maximum size in bytes taken by the modified transactions.                                   | 6            |
 
->**TODO**: Add the changes needed in LastCommitInfo for vote extensions
-
->**TODO**: DISCUSS: We need to make clear whether a proposer is also running the logic of a non-proposer node (in particular "ProcessProposal")
+>**TODO**: DISCUSS: We need to make clear whether a proposer is also running the logic of a non-proposer node (in particular "ProcessProposal").
```

Comment:
My sense is that this would be a bit a tricky, since most of the consensus code is a bit agnostic to if the node is the proposer of the current height/round. It treats all proposed blocks the same way, regardless of proposer. This could certainly be amended and bookkeeping could be implemented to track the needed state.

Reply:
Hmmm. Interesting.

From the spec's perspective, it's "simpler" to not distinguish between proposer and non-proposer. So everyone would receive the `ProcessProposal` call. If the proposer has already done the work as part of `PrepareProposal`, it can simply return "accept" to `ProcessProposal`. in the vast majority of Applications, this is expected to be the case.

The reason for this dilemma (and so, the TODO) is implementation/performance focused. Sending the whole proposed block over ABCI API is not a trivial thing to do in terms of performance if the Application is not built in the same process as Tendermint.

Now, the fact that the Tendermint logic currently doesn't distinguish between proposer and non-proposer is another variable to take into account; together with the fact that most Apps will be "in-process", so negligible performance penalty for unneeded `ProcessProposal` API calls.

After thinking about this for a while, I have a proposal, @williambanfield, @cmwaters, let me know what you think:

* For v0.36, we adopt the least disruptive way for the consensus core logic, and we don't distinguish between proposer and non-proposer. So the proposer's Application will also be called `ProcessProposal`.
* For v0.37 (or later, depending on priorities), we define a new Consensus param (or ".toml" param), which allows an Application to disable calling `ProcessProposal` on the proposer. The implementation of such parameter will be the added bookkeeping suggested by @williambanfield. This **option** will help with the performance of those Applications producing big blocks, which are not built in-process (although the _help_ wouldn't be huge, as only the proposer would stop receiving the blocks over serialized ABCI).

Reply:
> The reason for this dilemma (and so, the TODO) is implementation/performance focused. Sending the whole proposed block over ABCI API is not a trivial thing to do in terms of performance if the Application is not built in the same process as Tendermint.

I'm personally not sure that the performance savings is worth it. Proto serialization is reasonably fast and this isn't something a node has to do simultaneously or even all that often. When Tendermint calls `ProcessProposal` on the proposer, it does so at the same time as `ProcessProposal` is being called on the other nodes as well because they're all at the same step in the algorithm. (in reality, it happens a bit earlier on the proposer because it's gossiping the block directly to itself). Everyone then has to move through the propose->prevote step together before the network can make progress, so letting the proposer be 'faster' here doesn't seem to confer much benefit. All this to say, I'm not sure yet that a performance improvement in this step is that meaningful, but maybe there is something I'm missing. 

> we define a new Consensus param (or ".toml" param), which allows an Application to disable calling ProcessProposal on the proposer

As far as adding new parameters for this kind of thing, I think that we should try to reduce the number of knobs unless they are incredibly meaningful to application developers. They add maintenance burden and confusion for new users of the software.

Reply:
OK, so do we agree that:
* for v0.36 `ProcessProposal` is always called on all validators, including the proposer 
* for later, we wait for ABCI++ to start being adopted and, if there is demand for the (potential) optimization, then we can write an ADR to see how we add the option... and to discuss -- in line with your comments above -- whether it makes sense at all. That ADR could be reviewed by those that may ask for the optimization.

Reply:
Let me give this a quick look, my worry was that there may be more complexity related to reproposal, but perhaps that's not a problem.

Reply:
Looked at this together with @cmwaters. We can indeed call `isProposer` quite easily at that part of the code.
We have given this a spin: see this [patch](../commit/a62e098)
I've run "make build" and "make test", and everything is passing.
As a result, I'd align with @cmwaters and add this "if" in the code, which isn't complex and makes more sense for the Application; and keep the spec as it is (by removing this TODO; the spec is already aligned to not calling `ProcessProposal` at the proposer).

Reply:
Hmmm, actually, the condition in the patch is the opposite of what we want :blush:
But the point of the patch was to show it can be done easily.

Reply:
My secondary concern here was about the creation of candidate state. If an application relied on `ProcessProposal` for the creation of candidate state, then it wouldn't be run on proposers that re-propose locked blocks. As long as we make clear to application developers that they may not have candidate state in `FinalizeBlock` because these methods may not have been called, then there shouldn't be any danger in foregoing the call to `ProcessProposal` on the proposer, even if the proposer is re-proposing its locked block.

Reply:
I see where you are trying to get at here. My answer to that has two parts:
* As you say above, the Application can _never_ assume that it will always have a candidate state ready for a block delivered by `FinalizeBlock`. An Application trying to enforce that, will need an unbounded amount of memory just to keep all candidate states.
* In my understanding, a locked value in a proposed got to that state (a) either because that proposer proposed it in a previous round, or (b) because it received the proposal _and_ 2/3+ prevotes from other validators. In case (a) the Application received the proposal via `PrepareProposal`previously. In case (b) the Application received it via `ProcessProposal` in a previous round.